lib/protobuf/field_options_processor.ex

```diff
@@ -0,0 +1,103 @@
+defmodule Protobuf.FieldOptionsProcessor do
+  @moduledoc """
+  Defines hooks to process custom field options.
+  """
+
+  @typedoc """
+  Keyword list of field options. Right now only [extype: mytype].
+  """
+  @type options :: Keyword.t(String.t)
+
+  @typedoc """
+  The existing type of the field. Often the module name of the struct.
+  """
+  @type type :: atom
+
+  @typedoc """
+  A value with type extype.
+  """
+  @type value :: struct | any
+  @callback type_to_spec(type_enum :: atom, type :: String.t(), repeated :: boolean, options) :: String.t()
+  @callback type_default(type, options) :: any
+  @callback new(type, value, options) :: value
+  @callback encode_type(type, value, options) :: binary
+  @callback decode_type(val :: binary, type, options) :: value
+
+  @wrappers [
+    Google.Protobuf.DoubleValue,
+    Google.Protobuf.FloatValue,
+    Google.Protobuf.Int64Value,
+    Google.Protobuf.UInt64Value,
+    Google.Protobuf.Int32Value,
+    Google.Protobuf.UInt32Value,
+    Google.Protobuf.BoolValue,
+    Google.Protobuf.StringValue,
+    Google.Protobuf.BytesValue
+  ]
+
+  @wrappers_str [
+    "Google.Protobuf.DoubleValue",
+    "Google.Protobuf.FloatValue",
+    "Google.Protobuf.Int64Value",
+    "Google.Protobuf.UInt64Value",
+    "Google.Protobuf.Int32Value",
+    "Google.Protobuf.UInt32Value",
+    "Google.Protobuf.BoolValue",
+    "Google.Protobuf.StringValue",
+    "Google.Protobuf.BytesValue"
+  ]
+
+  def get_extype_mod(type) do
+    cond do
+      type in @wrappers -> Protobuf.Extype.Wrappers
```

Comment:
is it possible to structure the code in a way that it easier to add a new `Protobuf.Extype`?

Reply:
I'm also concerned that this current implementation ties up to closely with the current `elixir-protobuf` library. I do believe that the`elixir-protobuf` library needs to handle extensions, but the actual registration and implementation of the extensions should ideally leave somewhere else.

Is it something that's being discussed w/ Tony?

Reply:
I'm also concerned. I'm SO concerned. I don't like that the extension & implementation lives here. Here's why I chose to do it this way: 

1) If we want generated code to show presence of an extension like `field ... my field options: [extype:String.t]` then the extension (eg brex.elixirpb.ex) needs to be loaded https://github.com/tony612/protobuf-elixir/blob/e1d4a23baa1f51c2da58db24f58a26dc4d6511b2/lib/application.ex#L29 when protobuf-elixir is being run. However, normally we don't have any of our apps running when we run `protoc --plugin=protoc-gen-elixir ...`. So when I had brex.elixirpb.ex in our codebase it wasn't being picked up. I thought it was valuable for the generated code to show the extension is there so you know the extension has been picked up and is working correctly. (I had this hack where I could manually inject the extension  https://github.com/brexhq/credit_card/blob/0d53fd8c579fe11751c0dca06c3dbfbdfb59386a/libraries/brex_protobuf/lib/gather_extypes.ex#L53 but the generated code didn't show it.)

2) I like the idea of pluggable modules, like how Plug does, but it wasn't very compatible with protobuf-elixir being an escript.

Background: To use protobuf-elixir you generate an executable called an escript and then install it right. The installed escript is use to generate the pb.ex file and then the protobuf-elixir dependency you have in your app (like pulled in by the mix file`{:protobuf ~>"0.6"}`) is used to evaluate the pb.ex file and expand the `Protobuf` using module.

I considered passing a module through cmd line args like so https://github.com/brexhq/protobuf-elixir/blob/32945181e653a7e20dd8836bfd418ea04b01ec7c/lib/protobuf/protoc/cli.ex#L75. It worked, but if that module was ever called by the escript during code generation then it would fail - bc the module in the compiled executable. This means the extension could never alter code generation. 

I rejected this approach bc a) I wanted to change the generated types & b) it felt very brittle to assume my module wouldn't be called during code generation - encode, decode, and new (which are all altered by the module I passed) are called multiple times in the generation of the pb.ex files.

////////////////////////

Obviously we could not touch protobuf-elixir or the generated pb.ex files at all and then implement something that is similar to jon jon's caster or roll our own encoding/decoding function & codec. I kind of did the former here in my preprocessing pass strategy https://github.com/brexhq/credit_card/blob/851d16cab84d9440d3622cd94e197b2d4580ed9a/libraries/brex_protobuf/lib/build_mapping.ex 

We could also be smarter and address 1) by telling protobuf-elixir to look at `hard coded path` to find our extension. And address 2) by better design that ensures the module I pass won't be called or has a fallback or something. But both seem brittle/hacky to me. 😢 

I'm stressed. I guess the question is will merging this and using it for now not "preserve optionality"? Will we get ourselves into a state where we have multiple types of pb.ex files flying around that are incompatible and evaluated differently in different contexts 😱?!

I think if we keep the end user behaviour the same through iterations then we should be ok. I can do batch regeneration of pb.ex files if I have to. But I am scared.

What do you think?

Reply:
0. I tend to think that we should move off casters towards the generated code. We don't like much value we are getting is explicit tagging of fields.
1. I agree we need to approach the work iteratively;
2. I think it's unlikely that we need to regenerate pb.ex files because most of the logic is in macros. If you think there more code can be moved off pb.ex we should consider doing it.
3. It is another reason to automate pb.ex generation :)
4. `protobuf-elixir` generates pb.ex files that are compiled later so there must be a way to make it work. I wouldn't about it right now.